source/module_basis/module_nao/hydrogen_radials.cpp

```diff
@@ -0,0 +1,412 @@
+                             std::vector<double>& rvalue,
+                             const double sigma)
+{
+    double prefactor = 1.0 / sqrt(2.0 * M_PI) / sigma;
+    double rmax = rgrid.back();
+    for(int ir = 0; ir != rgrid.size(); ++ir)
+    {
+        double delta_r = rgrid[ir] - rmax;
+        double smooth = prefactor * exp(-delta_r * delta_r / 2.0 / sigma / sigma);
+        rvalue[ir] *= (1 - smooth);
+    }
+}
+
+std::map<std::pair<int, int>, std::pair<std::vector<double>, std::vector<double>>> 
+HydrogenRadials::generate_orb(const double charge,
+                              const int nmax,
+                              const double dr,
+                              const double conv_thr,
+                              const int rank,
+                              const std::string strategy,
+                              std::ofstream* ptr_log)
+{
+    // create space for storing all generated orbitals
+    // (n, l) to (rgrid, rvalue)
+    std::map<std::pair<int, int>, std::pair<std::vector<double>, std::vector<double>>> radials; 
+    std::map<std::pair<int, int>, double> rmaxs;
+
+    std::vector<std::pair<int, int>> nl_pairs = unzip_strategy(nmax, strategy);
+    double rmax = 0.0;
+    for(auto nl_pair : nl_pairs)
+    {
+        int n = nl_pair.first;
+        int l = nl_pair.second;
+        std::vector<double> rgrid;
+        std::vector<double> rvalue;
+        double rmax_nl = generate_hydrogen_radial_toconv(charge,
+                                                         n,
+                                                         l,
+                                                         conv_thr,
+                                                         rank,
+                                                         rgrid,
+                                                         rvalue,
+                                                         ptr_log);
+        radials[nl_pair] = std::make_pair(rgrid, rvalue);
+        rmaxs[nl_pair] = rmax_nl;
+        if(rmax < rmax_nl)
+        {
+            rmax = rmax_nl;
+        }
+    }
+    // zero padding towards rmax
+    for(auto& radial : radials)
+    {
+        int n = radial.first.first;
+        int l = radial.first.second;
+        std::pair<int, int> nl_pair = std::make_pair(n, l);
+
+        std::vector<double>& rgrid = radial.second.first;
+        std::vector<double>& rvalue = radial.second.second;
+
+        if(rmaxs[nl_pair] < rmax)
+        {
+            int ngrid = static_cast<int>((rmax - rmaxs[nl_pair]) / dr); // (r, rmax]
+            for(int ir = 1; ir <= ngrid; ++ir)
+            {
+                rgrid.push_back(rmaxs[nl_pair] + ir * dr);
+                rvalue.push_back(0.0);
+            }
+        }
+        // smooth the tail
+        smooth(rgrid, rvalue, 0.1);
+    }
+    return radials;
+}
+
+std::map<std::pair<int, int>, std::pair<int, int>>
+HydrogenRadials::mapping_nl_lzeta(const int nmax,
```

Comment:
initialy the orbitals are indiced by pair (l, zeta) instead of (n,l), so map between them two